lib/std/mem.zig

```diff
@@ -630,6 +631,122 @@ test "spanZ" {
+pub fn scanZ(ptr: anytype) Span(@TypeOf(ptr)) {
+    if (@typeInfo(@TypeOf(ptr)) == .Optional) {
+        if (ptr) |non_null| {
+            return scanZ(non_null);
+        } else {
+            return null;
+        }
+    }
+    const Result = Span(@TypeOf(ptr));
+    if (@typeInfo(Result).Pointer.sentinel) |s| {
+        if (s != 0) {
+            @compileError("std.mem.scanZ requires the sentinel, if any, to be 0");
+        }
+        return ptr[0..scanLenZ(ptr) :s];
+    } else {
+        return ptr[0..scanLenZ(ptr)];
+    }
+}
+
+test "scanZ" {
+    testing.expectEqualSlices(u8, "aoeu", scanZ("aoeu"));
+
+    {
+        var array: [5]u16 = [_]u16{ 1, 2, 3, 4, 5 };
+        testing.expectEqualSlices(u16, &array, scanZ(&array));
+        testing.expectEqualSlices(u16, array[0..3], scanZ(array[0..3]));
+        array[2] = 0;
+        testing.expectEqualSlices(u16, array[0..2], scanZ(&array));
+        testing.expectEqualSlices(u16, array[0..2], scanZ(array[0..3]));
+        const sentinel_ptr = @as([*:0]u16, array[0..2 :0]);
+        testing.expectEqualSlices(u16, array[0..2], scanZ(sentinel_ptr));
+        const opt_sentinel_ptr = @as(?[*:0]u16, array[0..2 :0]);
+        testing.expectEqualSlices(u16, array[0..2], scanZ(opt_sentinel_ptr).?);
+        const c_ptr = @as([*c]u16, &array);
+        testing.expectEqualSlices(u16, array[0..2], scanZ(c_ptr));
+        const slice: []u16 = &array;
+        testing.expectEqualSlices(u16, array[0..2], scanZ(slice));
+        const sentinel_slice: [:0]u16 = array[0..2 :0];
+        testing.expectEqualSlices(u16, array[0..2], scanZ(sentinel_slice));
+    }
+    {
+        var array: [5:0]u16 = [_:0]u16{ 1, 2, 3, 4, 5 };
+        testing.expectEqualSlices(u16, &array, scanZ(&array));
+        array[2] = 0;
+        testing.expectEqualSlices(u16, array[0..2], scanZ(&array));
+    }
+
+    testing.expectEqual(@as(?[]u8, null), scanZ(@as(?[]u8, null)));
+}
+
+/// Private helper for scanZ. If you want the scanned length, use scanZ(foo).len
+fn scanLenZ(ptr: anytype) usize {
+    switch (@typeInfo(@TypeOf(ptr))) {
+        .Pointer => |info| switch (info.size) {
+            .One => switch (@typeInfo(info.child)) {
+                .Array => |x| {
+                    if (x.sentinel) |sentinel| {
+                        assert(sentinel == 0);
+                        return indexOfSentinel(x.child, sentinel, ptr);
+                    } else {
+                        return indexOfScalar(x.child, ptr, 0) orelse x.len;
+                    }
+                },
+                else => {},
+            },
+            .Many => if (info.sentinel) |sentinel| {
+                assert(sentinel == 0);
+                return indexOfSentinel(info.child, sentinel, ptr);
+            },
+            .C => {
+                assert(ptr != null);
+                return indexOfSentinel(info.child, 0, ptr);
+            },
+            .Slice => if (info.sentinel) |sentinel| {
+                assert(sentinel == 0);
```

Comment:
why? there's no reason to ban other sentinels.

Reply:
When no sentinel is present in the type, this function assumes that the sentinel is zero.  So it would be a bit weird to both assume zero and allow nonzero.